drivers/dma/dma_silabs_ldma.c

```diff
@@ -0,0 +1,548 @@
+/*
+ * Copyright (c) 2024 Silicon Laboratories Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+#include <zephyr/sys/mem_blocks.h>
+
+#include "em_ldma.h"
+
+#define DT_DRV_COMPAT silabs_ldma
+
+#define DMA_IRQ_PRIORITY 3
+
+LOG_MODULE_REGISTER(silabs_dma, CONFIG_DMA_LOG_LEVEL);
+
+struct dma_silabs_channel {
+	enum dma_channel_direction dir;
+	uint32_t complete_callback_en;
+	atomic_t busy;
+	void *user_data;
+	dma_callback_t cb;
+	LDMA_TransferCfg_t xfer_config;
+	LDMA_Descriptor_t *desc;
+};
+
+struct dma_silabs_config {
+	void (*config_irq)(const struct device *dev);
+	const struct device *clock_dev;
+};
+
+struct dma_silabs_data {
+	struct dma_context dma_ctx;
+	struct dma_silabs_channel *dma_chan_table;
+	struct sys_mem_blocks *dma_desc_pool;
+};
+
+struct table {
+	int native;
+	int efr;
+};
+
+const static struct table ldma_blocksize_map[] = {
+	{0x1, ldmaCtrlBlockSizeUnit1},     {0x2, ldmaCtrlBlockSizeUnit2},
+	{0x3, ldmaCtrlBlockSizeUnit3},     {0x4, ldmaCtrlBlockSizeUnit4},
+	{0x6, ldmaCtrlBlockSizeUnit6},     {0x8, ldmaCtrlBlockSizeUnit8},
+	{0x10, ldmaCtrlBlockSizeUnit16},   {0x20, ldmaCtrlBlockSizeUnit32},
+	{0x40, ldmaCtrlBlockSizeUnit64},   {0x80, ldmaCtrlBlockSizeUnit128},
+	{0x100, ldmaCtrlBlockSizeUnit256}, {0x200, ldmaCtrlBlockSizeUnit512},
+	{0x400, ldmaCtrlBlockSizeUnit1024}
+};
+
+static int dma_silabs_get_transfer_parameter(uint32_t src_dsize, uint32_t dst_dsize,
+					     uint32_t block_size, LDMA_Descriptor_t *desc)
+{
+	if (src_dsize != dst_dsize) {
+		LOG_ERR("Source data size(%u) and destination data size(%u) must be equal",
+			src_dsize, dst_dsize);
+		return -ENOTSUP;
+	}
+
+	desc->xfer.size = LOG2(src_dsize);
+	if (desc->xfer.size < ldmaCtrlSizeByte || desc->xfer.size > ldmaCtrlSizeWord) {
+		return -ENOTSUP;
+	}
+
+	if (block_size % src_dsize) {
+		desc->xfer.xferCnt = block_size / src_dsize;
+	} else {
+		desc->xfer.xferCnt = block_size / src_dsize - 1;
+	}
+
+	if (desc->xfer.xferCnt > LDMA_DESCRIPTOR_MAX_XFER_SIZE) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int dma_silabs_get_blocksize(uint32_t src_blen, uint32_t dst_blen, uint32_t src_dsize)
```

Comment:
`dma_silabs_get_transfer_parameter()` fill the descriptor while `dma_silabs_get_blocksize()` returns a value. I suggest to align the APIs.